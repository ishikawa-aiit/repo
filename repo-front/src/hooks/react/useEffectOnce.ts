import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
