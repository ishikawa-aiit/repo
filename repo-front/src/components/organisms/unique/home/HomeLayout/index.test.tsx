import { screen, render, within } from '@testing-library/react'

import { HomeLayout } from '.'

const contents = 'サンプルテキスト'

describe('props:children', () => {
  it('正常にレンダリングされている', () => {
    const { container } = render(<HomeLayout>{contents}</HomeLayout>)
    const { getByText } = within(screen.getByTestId('HomeLayout'))

    expect(getByText(contents)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
