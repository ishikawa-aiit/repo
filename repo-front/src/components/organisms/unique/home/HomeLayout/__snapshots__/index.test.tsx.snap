// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props:children 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-height: 100vh;
  margin-top: 16px;
}

.emotion-1 {
  width: 100%;
  height: 100%;
  overflow: visible;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: neutral.600;
  width: 100%;
  margin-top: auto;
  font-size: sm;
  padding-top: 3px;
  padding-bottom: 3px;
  border-top: 1px;
  border-top-color: neutral.200;
}

<div>
  <div
    class="emotion-0"
    data-cy="HomeLayout"
    data-testid="HomeLayout"
  >
    <main
      class="emotion-1"
      data-cy="Main"
      data-testid="Main"
    >
      サンプルテキスト
    </main>
    <footer
      class="emotion-2"
      data-cy="Footer"
      data-testid="Footer"
    >
      © Advanced Institute of Industrial Technology, Oikawa Project Team 2022
    </footer>
  </div>
</div>
`;
